### PR TITLE
fix(tests): always tear down FDB on test failure

### PR DESCRIPTION
**Pull request by lgsilva3087**
Route ERR trap through test_end so teardown runs after failures. Harden FDB cleanup with PID-based stop, targeted pkill, wait loops, and worker process cleanup before removing the test workspace. Use is_fdb_cluster_running guard instead of global fdbmonitor pgrep.

### CODE REVIEW

File: tests/tools/foundationdb.sh [hunk 1]
```diff
@@ -1,6 +1,24 @@
...
+		if (( SECONDS >= deadline )); then
+			return 1
+		fi
+		sleep 0.1
+	done
+	return 0
+}
+
+function is_fdb_cluster_running() {
+	[[ -d "${workspace}" || -f "${workspace}/fdbmonitor.pid" ]]
```
[the remark this answers is not in the document]
Comment (lgsilva3087) [reply tests/tools/foundationdb.sh]: Fixed. is_fdb_cluster_running now uses pgrep -f "${fdbmonitor_pattern}" instead of checking for directory/pid file existence. This correctly returns true only when the fdbmonitor process for this specific test workspace is actually running.

File: tests/tools/foundationdb.sh [hunk 2]
```diff
@@ -41,8 +59,28 @@ function check_cluster_status() {
...
-	sudo pkill -f fdbmonitor || true
-	rm -r "${workspace:?}" || true
+	local pid_file="${workspace}/fdbmonitor.pid"
+
+	# First try graceful monitor shutdown by PID if available.
+	if [[ -f "${pid_file}" ]]; then
+		local monitor_pid
+		monitor_pid="$(cat "${pid_file}" 2>/dev/null || true)"
+		if [[ -n "${monitor_pid}" ]]; then
+			sudo kill "${monitor_pid}" 2>/dev/null || true
```
[the remark this answers is not in the document]
Comment (lgsilva3087) [reply tests/tools/foundationdb.sh]: Fixed. Before sending sudo kill, we now validate three conditions: the PID is numeric (=~ ^[0-9]+$), the process is still alive (kill -0), and the PID actually belongs to our fdbmonitor by cross-referencing with pgrep -f "${fdbmonitor_pattern}" | grep -qx. All three must pass, otherwise the kill is skipped and the pattern-based fallback handles it.

File: tests/tools/foundationdb.sh [hunk 3]
```diff
@@ -41,8 +59,28 @@ function check_cluster_status() {
...
+	sudo pkill -f "${fdbmonitor_pattern}" 2>/dev/null || true
+	wait_until_processes_stop "${fdbmonitor_pattern}" 5 || true
+
+	# Ensure worker processes bound to the test workspace are not left alive.
+	pkill -f "fdbserver.*${workspace}/data/" 2>/dev/null || true
+	pkill -f "backup_agent.*${workspace}/logs/" 2>/dev/null || true
+	wait_until_processes_stop "fdbserver.*${workspace}/data/" 2 || true
+	wait_until_processes_stop "backup_agent.*${workspace}/logs/" 2 || true
+
+	rm -rf "${workspace:?}" 2>/dev/null || true
```
[the remark this answers is not in the document]
Comment (lgsilva3087) [reply tests/tools/foundationdb.sh:79]: Partially addressed. rm -rf still runs unconditionally (by design — we don't want stale workspaces left behind), but the SIGKILL escalation added in fixes 3 and 4 means processes are very unlikely to still be accessing the workspace when it's removed. A best-effort final removal is the correct semantics for test cleanup.

File: tests/tools/foundationdb.sh [hunk 4]
```diff
@@ -41,8 +59,28 @@ function check_cluster_status() {
...
+		local monitor_pid
+		monitor_pid="$(cat "${pid_file}" 2>/dev/null || true)"
+		if [[ -n "${monitor_pid}" ]]; then
+			sudo kill "${monitor_pid}" 2>/dev/null || true
+		fi
+	fi
+
+	# Fallback kill by exact test-cluster command line.
+	sudo pkill -f "${fdbmonitor_pattern}" 2>/dev/null || true
+	wait_until_processes_stop "${fdbmonitor_pattern}" 5 || true
```
[the remark this answers is not in the document]
Comment (lgsilva3087) [reply tests/tools/foundationdb.sh]: Fixed. The || true fallback is replaced with an explicit check: if wait_until_processes_stop times out after 5 seconds, we escalate to sudo pkill -9 -f "${fdbmonitor_pattern}" and wait up to 2 more seconds for it to stop.

File: tests/tools/foundationdb.sh [hunk 5]
```diff
@@ -41,8 +59,28 @@ function check_cluster_status() {
...
+
+	# Ensure worker processes bound to the test workspace are not left alive.
+	pkill -f "fdbserver.*${workspace}/data/" 2>/dev/null || true
+	pkill -f "backup_agent.*${workspace}/logs/" 2>/dev/null || true
+	wait_until_processes_stop "fdbserver.*${workspace}/data/" 2 || true
+	wait_until_processes_stop "backup_agent.*${workspace}/logs/" 2 || true
```
[the remark this answers is not in the document]
Comment (lgsilva3087) [reply tests/tools/foundationdb.sh]: Fixed. Worker patterns are now in an array. The first loop sends SIGTERM to all workers in parallel. The second loop checks each one individually and escalates to pkill -9 if it hasn't stopped within the grace period, followed by a final 1-second wait.

File: tests/tools/foundationdb.sh [hunk 6]
```diff
@@ -41,8 +59,42 @@ function check_cluster_status() {
...
+		fi
+	fi
+
+	# Fallback kill by exact test-cluster command line; escalate to SIGKILL
+	# if the monitor does not stop within the grace period.
+	sudo pkill -f "${fdbmonitor_pattern}" 2>/dev/null || true
+	if ! wait_until_processes_stop "${fdbmonitor_pattern}" 5; then
+		sudo pkill -9 -f "${fdbmonitor_pattern}" 2>/dev/null || true
+		wait_until_processes_stop "${fdbmonitor_pattern}" 2 || true
+	fi
```
[the remark this answers is not in the document]
Comment (lgsilva3087) [reply tests/tools/foundationdb.sh]: Fixed. The PID-based sudo kill block was removed entirely — the sudo pkill -f path already handles graceful shutdown and is safer (no stale PID risk). For SIGKILL escalation, sudo pkill -9 -f fdbmonitor* was added to the sudoers configuration in setup_machine.sh, so the escalation path is now explicitly permitted in CI.

File: tests/tools/foundationdb.sh [hunk 7]
```diff
@@ -1,6 +1,24 @@
 #!/usr/bin/env bash
 
 readonly workspace="/tmp/saunafs-fdb-test"
+readonly fdbmonitor_pattern="fdbmonitor --conffile ${workspace}/conf/foundationdb.conf"
```
[the remark this answers is not in the document]
Comment (lgsilva3087) [reply tests/tools/foundationdb.sh]: Fixed. The dot in foundationdb.conf is now escaped as foundationdb\.conf in fdbmonitor_pattern, so pgrep/pkill -f treat it as a literal character rather than a regex wildcard.

File: tests/tools/foundationdb.sh [hunk 8]
```diff
@@ -41,8 +59,28 @@ function check_cluster_status() {
...
+	# Ensure worker processes bound to the test workspace are not left alive;
+	# escalate to SIGKILL for each worker pattern on timeout.
+	local worker_patterns=("fdbserver.*${workspace}/data/" "backup_agent.*${workspace}/logs/")
+	for pattern in "${worker_patterns[@]}"; do
+		pkill -f "${pattern}" 2>/dev/null || true
+	done
+	for pattern in "${worker_patterns[@]}"; do
+		if ! wait_until_processes_stop "${pattern}" 2; then
+			pkill -9 -f "${pattern}" 2>/dev/null || true
+			wait_until_processes_stop "${pattern}" 1 || true
```
[the remark this answers is not in the document]
Comment (lgsilva3087) [reply tests/tools/foundationdb.sh]: Fixed. The trailing slash was removed from the backup_agent pattern. The config sets logdir = ${workspace}/logs without a trailing slash, so the previous pattern would never match the actual backup_agent command line.

File: tests/tools/test.sh [hunk 9]
```diff
@@ -69,8 +69,8 @@ test_end() {
 		remove_all_emulated_zoned_disks
 	fi
 
-	# Stop FDB cluster if fdbmonitor is running
-	if pgrep fdbmonitor > /dev/null 2>&1; then
+	# Stop FDB cluster if a test-local FDB monitor/worker is running.
+	if test_function_exists is_fdb_cluster_running && is_fdb_cluster_running; then
```
[the remark this answers is not in the document]
Comment (lgsilva3087) [reply tests/tools/test.sh]: Addressed by redesign. Both is_fdb_cluster_running and stop_fdb_cluster have been removed entirely. test_end now checks the fdb_cluster_started flag (set by start_fdb_cluster) and calls cleanup_fdb_cluster directly. This eliminates all false negatives (orphaned workers, crashed fdbmonitor) and false positives (non-FDB tests triggering unnecessary pgrep calls). cleanup_fdb_cluster is idempotent so it handles all edge cases safely.

File: tests/tools/test.sh [hunk 10]
```diff
@@ -128,7 +128,7 @@ test_begin() {
 	test_cleanup
 	remove_all_emulated_zoned_disks
 	touch "$test_result_file"
-	trap 'trap - ERR; set +eE; catch_error_ "${BASH_SOURCE:-}" "${LINENO:-}" "${FUNCNAME:-}"; exit 1' ERR
+	trap 'trap - ERR; set +eE; catch_error_ "${BASH_SOURCE:-}" "${LINENO:-}" "${FUNCNAME:-}"; test_end' ERR
```
[the remark this answers is not in the document]
Comment (lgsilva3087) [reply tests/tools/test.sh:131]: Not a valid concern. test_freeze_result is intentionally called before cleanup commands that might fail (e.g. killall). If a test passed all assertions and froze the result, a subsequent ERR during cleanup should not flip the result to failure — that was a bug in the old behavior (exit 1 on any ERR, even post-freeze). When ERR fires before freeze, catch_error_ calls test_add_failure which writes to test_result_file normally (since test_frozen() is false), and test_end correctly exits 1. The new behavior is strictly more correct in both cases.

File: tests/setup_machine.sh [hunk 11]
```diff
@@ -191,6 +191,7 @@ if [ ! -f /etc/sudoers.d/saunafstest ] || ! grep -q '# FoundationDB' /etc/sudoer
 		# FoundationDB
 		saunafstest ALL = NOPASSWD: /usr/lib/foundationdb/fdbmonitor --conffile*
 		saunafstest ALL = NOPASSWD: /usr/bin/pkill -f fdbmonitor*
+		saunafstest ALL = NOPASSWD: /usr/bin/pkill -9 -f fdbmonitor*
 	END
 fi
```
[the remark this answers is not in the document]
Comment (lgsilva3087) [reply tests/setup_machine.sh:197]: Out of scope.

**Review by ralcolea — APPROVED** [review]
Nice job! 💪 🔥 🚀

**Review by dmga44 — APPROVED** [review]
LGTM.
